rd-net/RdFramework/Impl/StealingScheduler.cs

```diff
@@ -0,0 +1,134 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+
+namespace JetBrains.Rd.Impl
+{
+  /// <summary>
+  /// A scheduler with a ability to join all queued tasks at some point and execute them synchronously when necessary
+  /// </summary>
+  public class StealingScheduler : TaskScheduler
+  {
+    private readonly bool myAllowParallelProcessing;
+    [NotNull] private readonly TaskScheduler myScheduler;
+
+    /// <summary>
+    /// >0: The number of currently simultaneously running tasks.
+    /// -1: exclusive mode taken by Join when <see cref="myAllowParallelProcessing"/> is false, no tasks allowed to process 
+    /// </summary>
+    private volatile int myActive;
+    [NotNull] private readonly ConcurrentQueue<Task> myActions = new ConcurrentQueue<Task>();
+
+    /// <summary>
+    /// Creates stealing scheduler
+    /// </summary>
+    /// <param name="scheduler">Scheduler which used to queue actions in <see cref="ConcurrentQueue"/></param>
+    /// <param name="allowParallelProcessing">
+    /// Indicates whether is it safe to execute tasks simultaneously in <see cref="Join"/> method and provided scheduler.
+    /// Note, that this is your responsibility to provide limited concurrency scheduler when concurrent processing is forbidden.
+    /// </param>
+    public StealingScheduler([CanBeNull] TaskScheduler scheduler, bool allowParallelProcessing = true)
+    {
+      myAllowParallelProcessing = allowParallelProcessing;
+      myScheduler = scheduler ?? Default;
+    }
+
+    protected override IEnumerable<Task> GetScheduledTasks()
+    {
+      return myActions;
+    }
+
+    protected override void QueueTask(Task task)
+    {
+      myActions.Enqueue(task);
+      if (myActive < myScheduler.MaximumConcurrencyLevel)
```

Comment:
Current implementation could lead to temporary bursts of activity when there are more `ProcessTasks` tasks scheduled than `myScheduler.MaximumConcurrencyLevel` (because `myActive` only considers threads that process a task at the current moment of time, and threads executing `ProcessTasks` could be in other states, too).

Say, if the scheduler is currently occupied by something else, then `myActive` will stay at zero, and thus any call to `QueueTask` will allocate and schedule a fresh task.

What do you think? Could/should we set a harder limit on the count of these tasks (say, try to terminate `ExecuteOne` in case of `myActive >= myScheduler.MaximumConcurrencyLevel`)? Or was it done deliberately the way it is?

Reply:
The point of this check is not to strictly follow MaximumConcurrencyLevel of the wrapped scheduler, but rather just to avoid creating and queuing tasks on `myScheduler` when it is definitely pointless. We aware of the "proccess all" behavior of `StealingScheduler` and that is why it is valid to apply this small optimization.
Strictly enforcing the amount of concurrent threads is `myScheduler` responsibility.

Reply:
I'm not talking about cases when several tasks run concurrently (since it isn't a production case even).

Consider a case when the scheduler is occupied by some long task, and there are several clients who want to do `QueueTask` (say, to register a new serializer). In this case, each client will cause the creation of a `new Task(ProcessTasks)`.

Later, after the scheduler is free, it will start processing these tasks (most likely, one by one), so the first task will do all the work, and others will do nothing, and will be GC'd later.

As I see the current implementation, it could create some excessive garbage on serializer registration, in cases when it isn't required.

Yes, it shouldn't be too much in practice (considering how we use this scheduler and how many serializers we register in each run), but still: don't you think we could do better here?

This is a genuine question; I'm not sure we should improve that, but I'd like to discuss.

Other than that, I have no concerns about this approach, but I'd like to wait for other reviewers. The scheduler code is tricky, and it's possible I've missed some important points.

Reply:
Sorry for long answer, I've waited for a possible comments from other reviewers.

Thanks for clarification, I've understood the case, but cannot find a neat way to implement it without additional tracking of scheduled tasks. Also, the impact of such optimization is also debatable - task tracking isn't free.

In the current production code it seems as incredible rare scenario. We are using exclusive scheduler over default thread pool. This scheduler may be occupied for a long time when either thread pool starvation happens or exclusive scheduler already executing our task. In the second case we won't schedule new tasks because myActive == MaxConcurrencyLevel.